Main 타이머 애니메이션 구현

Minny27 opened the pull request:
## Motivation ⍰

- 타이머 애니메이션이 기본 형태가 아니라 커스텀해야 함

<br>

## Key Changes 🔑

- 미리 보기는 다음과 같습니다.

<img src="https://github.com/THT-Team/THT-iOS/assets/68800789/86182a87-1785-40cb-850c-2c5d89099359" width=25%>


- trackLayer(애니메이션의 배경 레이어), strokeLayer(색상이 변하는 레이어), dotLayer(점 레이어) 세 레이어로 나눔

|trackLayer|strokeLayer|dotLayer|
|:---:|:---:|:---:|
|<img src="https://github.com/THT-Team/THT-iOS/assets/68800789/0dc8d6bf-50fe-4adc-91d8-611a3b77829d" width=80%>|<img src="https://github.com/THT-Team/THT-iOS/assets/68800789/85e1b24f-694e-45fd-a879-a76b65042ed3" width=80%>|<img src="https://github.com/THT-Team/THT-iOS/assets/68800789/faaa5568-e44b-4f03-9565-4765bc7db62f" width=80%>|

- trackLayer, strokeLayer는 동일한 Path로 구현
- 초를 8단계로 나눔
```swift
// MainViewModel
enum TimeState {
  case initial // 7~8
  case high  // 6~7
  case middle // 3~6
  case low // 2~3
  case zero // 1~2
  case over // 0~1
  
  init(rawValue: Double) {
    switch rawValue {
    case 7.0..<8.0:
      self = .initial
    case 6.0..<7.0:
      self = .high
    case 3.0..<6.0:
      self = .middle
    case 2.0..<3.0:
      self = .low
    case 1.0..<2.0:
      self = .zero
    default:
      self = .over
    }
  }
  {...}
}
```
- Observable<Int>.interval를 통해 0~8초를 emit함
- 근데 프로그래스 애니메이션이 천천히 줄어들도록하기 위해 1초를 100개로 나눔
- 그래서 0 ~ 800까지를 emit함
```swift
// MainViewModel
func transform(input: Input) -> Output {
    let time = Observable<Int>.interval(.milliseconds(10),
                                        scheduler: MainScheduler.instance)
      .take(8 * 100 + 1)
      .map { round((8 - Double($0) / 100) * 100) / 100 }
      .asDriver(onErrorDriveWith: Driver<Double>.empty())
    {...}
}
```
- timer가 time을 emit할 때마다 dot의 Position을 갱신. 여기서 stokeLayer보다 먼저 나오도록 sine, cosine에 값을 더하는 방식으로 구현
```swift
// MainViewModel
func transform(input: Input) -> Output {
    {...}
    let dotPosition = progress.map { progress in
      var progress = round(progress * 100) / 100
      if progress > 0.95 || progress < -0.05 { progress = 0.95 }
      let radius = CGFloat(22 / 2 - 2 / 2)
      let angle = 2 * CGFloat.pi * CGFloat(progress) - CGFloat.pi / 2
      let dotX = radius * cos(angle + 0.3) // 값 추가
      let dotY = radius * sin(angle + 0.3) // 값 추가
      return CGPoint(x: dotX, y: dotY)
    }
    {...}
}
```

<br>

## To Reviewers 🙏🏻

- [ ] 디자인 가이드와 동일하게 동작하는지 확인해주세요
- 아직 start, stop은 구현하지 못했습니다.
- 점과 stroke layer와의 거리를 근사치로 해놨는데 더 좋은 방법이 있으면 제안해주세요
- 그리고 view model에서 dot position을 emit할 때 view의 bounds값이 필요한데 view의 bounds를 알기 위해서는 vc에서 처리해야되다보니 수치를 그대로 구현했는데 이 부분도 더 좋은 방법이 있을 지 궁금합니다.
-  그리고 radius를 구할 때 progress 값을 소수점 2번째 자리에서 반올림해야 원하는 대로 점이 찍히는데
progress bar에서 Progress를 표시할 때는 소수점 2번째 자리를 포함하고 있어야 느리게 표현할 수 있어서 아래 처럼처리하고 있는데 더 좋은 방법이 있을 지 궁금합니다.
```swift
// MainViewController
override func bindViewModel() {
    output.progress
      .map { round(CGFloat($0) * 100) / 100 }
      .drive(mainView.timerView.strokeLayer.rx.strokeEnd)
      .disposed(by: disposeBag)
    
    output.progress
      .map { CGFloat($0) }
      .drive(mainView.progressView.rx.progress)
      .disposed(by: disposeBag)
}
```

<br>

## Linked Issue 🔗

- [x] #11

Minny27 committed:
[#25]Feat: CardProgressView 커스텀

- draw를 override해서 setNeedsDisplay() 메서드로
- progress 값이 변경될 때마다 다시 그리는 방식으로 구현
- bar color는 값만 변경되서 setNeedsLayout() 업데이트
- 1초를 100개로 쪼개서 소수점 2번째 자리에서 반올림하고
- 그 값을 5로 나눠서(0~1사이 값으로 표현하기 위해) CGFloat으로 emit

Minny27 committed:
[#11]Feat: timer animation 구현

- trackLayer(애니메이션의 배경 레이어), strokeLayer(색상이 변하는 레이어), dotLayer(점 레이어) 세 레이어로 나눔
- strokeLayer에서는 progress에 따라 애니메이션이 적용됨
- dotLayer는 strokeLayer보다 앞서서 position(x,y)를 변경하는 방식으로 구현
- 초를 0~8까지로 emit하고 구간은 8개로 나눔
- progress가 천천히 가도록 구현하기 위해서 1초를 100개로 나눠서
- 총 801개가 emit됨
- clear 색상 추가

Minny27 committed:
[#11] Feat: dotPath start, end angle Stroke layer와 동일하게 변경

ibcylon commented on Falling/Sources/Feature/Main/MainViewModel.swift:
```diff
@@ -19,63 +94,67 @@ final class MainViewModel: ViewModelType {
+    let dotPosition = progress.map { progress in
+      var progress = round(progress * 100) / 100
+      if progress > 0.95 || progress < -0.05 { progress = 0.95 }
+      let radius = CGFloat(22 / 2 - 2 / 2)
```
이 식에 대해서 주석 달아주시면 좋을 것 같습니다!

Minny27 replied:
넵!
'~=' 연산자도 알아가네요

ibcylon commented:
- 그리고 view model에서 dot position을 emit할 때 view의 bounds값이 필요한데 view의 bounds를 알기 위해서는 vc에서 처리해야되다보니 수치를 그대로 구현했는데 이 부분도 더 좋은 방법이 있을 지 궁금합니다.
- 그리고 radius를 구할 때 progress 값을 소수점 2번째 자리에서 반올림해야 원하는 대로 점이 찍히는데
progress bar에서 Progress를 표시할 때는 소수점 2번째 자리를 포함하고 있어야 느리게 표현할 수 있어서 아래 처럼처리하고 있는데 더 좋은 방법이 있을 지 궁금합니다.

Timer와 프로그래스바를 들고 있는 컴포넌트를 만드시고
Rx extention을 이용하여 Binder를 만드시면 좋을 것 같습니다.
- 뷰모델에서 어떻게 구현하는지 감출 수 있고, 따라서 코드 가독성이 증가합니다.
- Output Stream 개수를 줄일 수 있음
- 컴포넌트에서 그리기 때문에 view bound등의 값을 바로 얻을 수 있음.

ibcylon committed:
Refactor: using loadView for mainView()

ibcylon commented on Falling/Sources/Feature/Main/MainView.swift:
```diff
@@ -19,9 +22,12 @@ final class MainView: TFBaseView {
-  override func makeUI() {
+
+  override func layoutSubviews() {
     self.backgroundColor = .systemGray
```
makeUI에서 backgroundColor값을 조정해도 변경되지 않아서 여기서 변경하였습니다.

Minny27 replied:
이거는 이유를 모르겠네요..

ibcylon committed:
Refactor: Rx binder with time state

ibcylon commented on Falling/Sources/Feature/Main/MainView.swift:
```diff
@@ -55,4 +55,47 @@ final class MainView: TFBaseView {
+extension Reactive where Base: MainView {
+  var timeState: Binder<MainViewModel.TimeState> {
+    return Binder(self.base) { (base, state) in
+      base.timerView.timerLabel.textColor = state.color.color
```
Time state enum에 associate value 추가해서
값 처리하게 하였습니다.

Rx Binder 생성하여 state에 따른 component값 변경하게 하였습니다.

dotPosition 로직 메인뷰 안에 넣었습니다.

추후에는 타이머와 프로그래스를 가진 컴포넌트를 만들고 거기에 옮기는 게 좋을 것 같습니다.

Minny27 replied:
Reactive 확장해서 구현할 수 있군요..
dotPosition 구하는 게 비즈니스 로직이라고 생각했는데
뷰의 bounds, center 값으로 계산해서 뷰에 표시하다보니 view에서 작성하는게 맞겠네요

ibcylon commented on Falling/Sources/Feature/Main/MainViewModel.swift:
```diff
@@ -12,27 +12,27 @@ import Foundation
-    case low // 2~3
-    case zero // 1~2
-    case over // 0~1
+    case initial(value: Double) // 7~8
```
associate value 추가

ibcylon reviewed (approved):
수고하셨습니다.

Minny27 commented on Falling/Sources/Feature/Main/MainViewController.swift:
```diff
@@ -34,15 +34,8 @@ final class MainViewController: TFBaseViewController {
-      $0.trailing.equalToSuperview()
-    }
+  override func loadView() {
+    self.view = mainView
```
아 이런 방법이 있었군요! 배워갑니다

Minny27 committed:
[#25]Enhance: 점과 stroke layer 간격 수정

- 디자인과 더 일치하도록 점과 stroke layer의 간격을 더 벌림
- 2~4초가 같은 색인줄 알았으나 다른색임
- TimeState case를 초별로 다 나눔
- 색상 추가
- dotposition에 대한 주석 추가

Minny27 commented on Falling/Sources/Feature/Main/MainView.swift:
```diff
@@ -55,19 +55,19 @@ final class MainView: TFBaseView {
-    if !(range ~= progress) { progress = 0.95 }
-    let radius = CGFloat(22 / 2 - 2 / 2)
+    var progress = progress
+    // progress가 -0.05미만 혹은 1이상은 점(dot)을 0초에 위치시키기 위함
```
dot이 0초 위치를 초과하는 progress 범위에 대한 설명 주석 추가했습니다!

Minny27 commented on Falling/Sources/Feature/Main/MainView.swift:
```diff
@@ -55,19 +55,19 @@ final class MainView: TFBaseView {
-    let dotY = radius * sin(angle + 0.3)
-
+    let dotX = radius * cos(angle + 0.35)
+    let dotY = radius * sin(angle + 0.35)
```
점의 간격이 좁은 거같아서 근사치로 수정했습니다

Minny27 commented on Falling/Sources/Feature/Main/MainViewModel.swift:
```diff
@@ -13,9 +13,11 @@ final class MainViewModel: ViewModelType {
+    case four(value: Double) // 5~6
+    case three(value: Double) // 4~5
+    case two(value: Double) // 3~4
+    case one(value: Double) // 2~3
```
2~5초 색상이 같다고 생각했는데 달라서 모든 초별로 케이스 나눴습니다!
